src/ParquetMsg.chpl

```diff
@@ -871,6 +925,520 @@ module ParquetMsg {
+        when ARROWFLOAT   do return c_ptrTo((e:(BSE(real(64), 1))).a[off]);
+        when ARROWBOOLEAN do return c_ptrTo((e:(BSE(bool    , 1))).a[off]);
+        when ARROWDECIMAL do return c_ptrTo((e:(BSE(real(64), 1))).a[off]);
+        otherwise do
+          throw new NotImplementedError("Unexpected column type while " +
+                                        "reading parquet file", L(), R(), M());
+      }
+    }
+
+    inline proc getPtr(e: GenSymEntry?, off: int): c_ptr(void) throws {
+      return getPtr(e, off, this.ty);
+    }
+
+    // postProcess is called per-entry to handle some special cases
+    // Engin: when I first started implementing, there were more of these
+    // special cases. I have eliminated all, but null handling. We could
+    // consider making integral arrays always floats if the user enabled null
+    // handling.
+    proc postProcess(e: shared GenSymEntry?): shared AbstractSymEntry throws {
+      return postProcess(e, ty, 0);
+    }
+
+    proc postProcess(e: shared GenSymEntry?, ty,
+                     colIdx): shared AbstractSymEntry throws {
+      type SSE = shared SymEntry(?);
+      select ty {
+        when ARROWINT64   do return _postProcess(e:(SSE(int(64) , 1)), colIdx);
+        when ARROWINT32   do return _postProcess(e:(SSE(int(64) , 1)), colIdx);
+        when ARROWUINT64  do return _postProcess(e:(SSE(uint(64), 1)), colIdx);
+        when ARROWUINT32  do return _postProcess(e:(SSE(uint(64), 1)), colIdx);
+        when ARROWDOUBLE  do return _postProcess(e:(SSE(real(64), 1)), colIdx);
+        when ARROWFLOAT   do return _postProcess(e:(SSE(real(64), 1)), colIdx);
+        when ARROWBOOLEAN do return _postProcess(e:(SSE(bool    , 1)), colIdx);
+        when ARROWDECIMAL do return _postProcess(e:(SSE(real(64), 1)), colIdx);
+        otherwise do
+          throw new NotImplementedError("Unexpected column type while " +
+                                        "reading parquet file", L(), R(), M());
+      }
+    }
+
+    // I wanted to name these `postProcess` but ran into a compiler bug
+    // TODO instead of `colIdx`, we can actually pass `whereNull`
+    proc _postProcess(se: SymEntry(?t), colIdx) throws where validIntType(t) ||
+                                                             validUIntType(t) ||
+                                                             t==bool {
+      const ref myWhereNull = _whereNull[colIdx];
+      if hasNonFloatNulls && (|| reduce myWhereNull) {
+        // if we have non-float nulls and there's at least one null
+        var floatEntry = createSymEntry(se.size, real);
+        floatEntry.a = (se.a):real;
+        ref fa = floatEntry.a;
+        [(t, f) in zip(myWhereNull, fa)] if t then f = nan;
+
+        return floatEntry: AbstractSymEntry;
+      }
+      else {
+        return se: AbstractSymEntry;
+      }
+    }
+
+    inline proc _postProcess(se: SymEntry(?t), colIdx) throws {
+      return se;
+    }
```

Comment:
consider having `_postProcess` also return `AbstractSymEntry` to prevent extra generics. I would suggest using an explicit return type of `AbstractSymEntry` for both, like `postProcess`

Reply:
Can you elaborate why you think the current implementation would lead to more generics?